tests/Microsoft.Identity.Test.Integration.netfx/HeadlessTests/ClientCredentialsTests.cs

```diff
@@ -0,0 +1,395 @@
+            string issuer,
+            string aud,
+            X509Certificate2 cert)
+        {
+            // no need to add exp, nbf as JsonWebTokenHandler will add them by default.
+            var claims = new Dictionary<string, object>()
+            {
+                { "aud", aud },
+                { "iss", issuer },
+                { "jti", Guid.NewGuid().ToString() },
+                { "sub", issuer }
+            };
+
+            var securityTokenDescriptor = new SecurityTokenDescriptor
+            {
+                Claims = claims,
+                SigningCredentials = new X509SigningCredentials(cert)
+            };
+
+            var handler = new JsonWebTokenHandler();
+            var signedClientAssertion = handler.CreateToken(securityTokenDescriptor);
+
+            return signedClientAssertion;
+        }
+
+       
+        private static string GetSignedClientAssertionManual(
+            string issuer, // client ID
+            string audience, // ${authority}/oauth2/v2.0/token for AAD or ${authority}/oauth2/token for ADFS
+            X509Certificate2 certificate)
+        {
+            const uint JwtToAadLifetimeInSeconds = 60 * 10; // Ten minutes
+            DateTime validFrom = DateTime.UtcNow;
+            var nbf = ConvertToTimeT(validFrom);
+            var exp = ConvertToTimeT(validFrom + TimeSpan.FromSeconds(JwtToAadLifetimeInSeconds));
+
+            var payload = new Dictionary<string, string>()
+            {
+                { "aud", audience },
+                { "iss", issuer },
+                { "jti", Guid.NewGuid().ToString() },
+                { "sub", issuer },
+                { "exp", exp.ToString(CultureInfo.InvariantCulture) },
+                { "nbf", nbf.ToString(CultureInfo.InvariantCulture) },
+                { "iat", nbf.ToString(CultureInfo.InvariantCulture) }
+            };
+
+            RSACng rsa = certificate.GetRSAPrivateKey() as RSACng;
+
+            //alg represents the desired signing algorithm, which is SHA-256 in this case
+            //kid represents the certificate thumbprint
+            var header = new Dictionary<string, string>()
+            {
+              { "alg", "RS256"},
+              { "kid",  certificate.Thumbprint},
+              { "typ", "JWT"},
+              { "x5t", Encode(certificate.GetCertHash())},
+            };
+
+            string token = Encode(
+                Encoding.UTF8.GetBytes(JObject.FromObject(header).ToString())) +
+                "." +
+                Encode(Encoding.UTF8.GetBytes(JObject.FromObject(payload).ToString()));
+
+            string signature = Encode(
+                rsa.SignData(
+                    Encoding.UTF8.GetBytes(token),
+                    HashAlgorithmName.SHA256,
+                    System.Security.Cryptography.RSASignaturePadding.Pkcs1));
+            return string.Concat(token, ".", signature);
+        }
+
+        private static string Encode(byte[] arg)
```

Comment:
Could use existing helper? https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/blob/d8e77fe072d4326027ed7f75c97179a03ba0897e/src/client/Microsoft.Identity.Client/Utils/Base64UrlHelpers.cs#L65